card_live_dashboard/layouts/figures.py

```diff
@@ -114,6 +122,210 @@ def totals_figure(data: CardLiveData, type_value: str, color_by_value: str) -> g
+        category_sets = totals_df.reset_index().groupby('filename')\
+                                        .agg(lambda x: tuple(x)).applymap(list)
+        category_sets = category_sets['categories']\
+                         .apply(lambda x: sorted(x)).sort_values().apply(tuple)
+        category_sets = category_sets.value_counts()
+
+        # convert to upset data
+        upset_data = upsetplot.from_memberships(category_sets.index,
+                                                category_sets.values)
+        upset_data = upsetplot.UpSet(upset_data,
+                                     sort_by='cardinality')
+
+        # get number of sets and categories into UpSet class
+        num_sets, num_categories = upset_data.intersections.reset_index().shape
+
+        # if the data is empty
+        if num_sets == 0 or num_categories == 0:
+            fig = EMPTY_FIGURE
+        # or if there is way too much data to plot
+        elif num_categories > 40:
+            fig = go.Figure(layout={
+                                'xaxis': {'visible': False},
+                                'yaxis': {'visible': False},
+                                'annotations': [{
+                                'text': f"{num_categories} total {title} "
+                                         "categories with current selection. "
+                                         "<br>Please subset input data "
+                                         "further to display set "
+                                         "intersections.",
+                                'xref': 'paper',
+                                'yref': 'paper',
+                                'showarrow': False,
+                                'font': {'size': 28}
+                                }]
+                            })
+
+        else:
+			# filter to top 25 sets by cardinality
+            if num_sets > 25:
+                truncated = True
+                upset_data.intersections = upset_data.intersections[:25]
+            else:
+                truncated = False
+
+            # get the category names in each set for hover annotation
+            set_category_annotations = []
+            for mask in upset_data.intersections.index:
+                categories_in_set = itertools.compress(\
+                                          upset_data.intersections.index.names,
+                                          mask)
+                categories_in_set = "<br>".join(sorted(categories_in_set))
+                set_category_annotations.append(categories_in_set)
+
+            set_intersections = upset_data.intersections.reset_index()\
+                                    .iloc[:, :-1].astype(int).T.iloc[::-1]
+
+            # make grid for set memberships
+            grid_x = []
+            grid_y = []
+            membership_x = []
+            membership_y = []
+            names = []
+
+            y = 0
+            for row in set_intersections.iterrows():
```

Comment:
I think you can use Python's `enumerate()` function here https://docs.python.org/3/library/functions.html#enumerate to avoid having to create a separate index variable to increment. That is:

```python
for (y,row) in enumerate(set_intersections.iterrows()):
```

Reply:
Good point! Done.